[PowerToysRun][OneNote] Improve the OneNote plugin

Odotocodot opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

This pull request improves the OneNote plugin by adding new features such as a notebook explorer and the ability to quickly create new notebooks, section groups, sections and pages. 

<p>
<img height="200"  alt="base menu" src="https://github.com/user-attachments/assets/bbc0c467-3596-4b38-8f66-66a00f5450c4" />
<img src="https://github.com/user-attachments/assets/fbcac715-9fa9-4fb9-a9cc-a8f793a2ca58" alt="notebook explorer" height="300"/>
<img height="300" alt="onenote_search" src="https://github.com/user-attachments/assets/cc01b77d-0331-4ad8-95e5-d7f5fbf6ab31" />

</p>

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #29348, #22153
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [x] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

**List of all new features:**
- View notes hierarchically (called notebook explorer)
	- This allows users to open all item types (not just pages) in OneNote from PTRun.
- Quickly create new notebooks, section groups, sections and pages.
- Search items by their title only.
- Search pages only in a specific notebook, section group or section.
- View recently edited pages.
- New icons (option to match PowerToys theme or use colored icons).
- The ability to see whether an item has unread changes.

**This adds new plugin specific keywords:**
| Keyword | Description  |
|--------------|------------------|
| `nb:\`  | Access the notebook explorer  |
| `rp:`  | Quickly view recent pages  |
| `*`  | Search items by title  | 
| `>`  | Search pages within an item (Only works in notebook explorer) |


**New settings:**
- Show an icon for items with unread changes.
- Show encrypted sections (even if they are unlocked)
- Show the recycle bin section group
- Create coloured icons for notebooks and sections
- COM Object timeout.
	- The delay before the COM object is removed from memory after user input. Creating a COM object is a slow process.
	- This setting is fine left as default, so can potential be removed if too confusing.
	- This is a work around given there is no api to know when the run window is opened or closed.


 
This also changes package dependencies, one of them is that LazyCache is not used so the lines mention below may not be needed any more.
https://github.com/microsoft/PowerToys/blob/809791da25fd8c43d689804617acfcf7f68cedca/src/modules/launcher/Wox.Plugin/Wox.Plugin.csproj#L34-L35 

Also as stated in #29348, this pull request is essentially porting non conflicting features from this OneNote [plugin](https://github.com/Odotocodot/Flow.Launcher.Plugin.OneNote) directly into PowerToys Run.


<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

Manually tested: 
- Creating new items
- Search capabilites
- Correct icons and text displayed

Odotocodot committed:
Add Odotocodot.OneNote.Linq package reference

Odotocodot committed:
Pull files from old branch

Odotocodot committed:
Add plugin icons

Odotocodot committed:
Refactor classes

Odotocodot committed:
Remove unused package references

Removed Interop.Microsoft.Office.Interop.OneNote
Remove ScipBe.Common.Office.OneNote

Odotocodot committed:
Move user facing strings to resources

Odotocodot committed:
Simplify result tooltips

Added the Humanizer package
Moved user facing strings to resources

Odotocodot committed:
Small code clean up

Odotocodot committed:
Remove unused package references

Removed LazyCache package

Odotocodot committed:
Update plugin.json

Odotocodot committed:
Add words to spell checker

Odotocodot committed:
Fix incorrect query text

Odotocodot committed:
Refactor creating new OneNote item results

Odotocodot commented:
@microsoft-github-policy-service agree

Odotocodot committed:
Update Directory.Packages.props

Fix Mages package version

jaimecbernardo reviewed (commented):
Hi, thank you for opening this PR. Was about to review, but I think you did a bad merge here. The changes in the PR are totalling 413 files: https://github.com/microsoft/PowerToys/pull/36813/files

![image](https://github.com/user-attachments/assets/54bc5966-dc88-40b3-9808-e1124661445d)

Can you please fix the branch so that it only shows the actual changes that this PR is bringing in? Please let me know if you need some help doing this.

Odotocodot commented:
Oh, something went awry during my rebase. Yes, please any help will be much appreciated.

Odotocodot commented:
I have fixed the branch now!

moooyo commented:
Thanks for your amazing work. @Odotocodot 

I‘m sorry for my late review and reply to u. We are preparing for our new Command Palette plugin which will be the successor of PowerToys Run. 

Would you consider to put effort on our new Command Palette plugin? Here's the doc: https://github.com/zadjii-msft/PowerToys/blob/main/src/modules/cmdpal/doc/command-pal-anatomy/command-palette-anatomy.md

Odotocodot commented:
Hi @moooyo, I see, thanks for the information.
I would be happy to port it to the new Command Palette plugin.

moooyo commented:
> Hi @moooyo given the full release of the Command Palette plugin, is it prefered to work on porting the plugin to Command Palette or should the focus be resolving the current conflicts for the PowerToys Run?

Sorry, I missed this PR.

Would you want to merge it in? If yes, please resolve the conflict and ping me. I will follow up it. sry again it's long time to wait.

If you want to port it into cmd pal, I'd like to help you.

moooyo commented:
> No worries @moooyo, thanks for the resposnse.
> 
> I would love port it to cmd pal and help will be much appreciated.
> 
> This merge to PTRun would still require a user documentation update given the new features, so it might be worth doing this for both PTRun, and cmd pal when the port is finished,

Fine. Please feel free to at me if it's ready to review. I will follow up it.

moooyo commented:
I suggest to keep in both side. @Odotocodot 

We don't need to ref unnecessary resource in cmdpal. It need to keep independence.

Odotocodot committed:
Update dev docs

Updated comments

zadjii-msft commented:
(I'm gonna move the CmdPal specific discussion over to #39444)

niels9001 commented:
@odotocodot @zadjii-msft Do we want to continue with this PR or (help) with migrating this to Command Palette (which I think would be the better option)?

I see there are some questions raised here: https://github.com/microsoft/PowerToys/issues/39444#issuecomment-2883843015

Odotocodot committed:
Remove ComObjectTimeout setting

Not needed.

Odotocodot committed:
Migrate to `LinqToOneNote`

Odotocodot.OneNote.Linq 1.1.0 -> LinqToOneNote 2.1.0
Its the same package under and new package id in nuget

Odotocodot committed:
Remove `Humanizer` dependency

Odotocodot committed:
Refactor `SearchManager`

Odotocodot committed:
Update `ResultCreator`

Remove unneeded result, and add subtitle.

Odotocodot committed:
Update dev docs

niels9001 committed:
Merge branch 'main' into feature/improve-onenote-plugin-2

niels9001 commented:
@Odotocodot Are you planning to ship your Command Palette extension to the Store and/or Winget still :)? If so, make sure you submit it to the [Extension Gallery](https://github.com/microsoft/CmdPal-Extensions/) too!

Edit: closing this PR for now. Feel free to re-open!
